File: lib/engines/technical.ts
import type { OHLCVBar } from "@/lib/data/types";
import { evaluateStrategyConfluence } from "./strategies";

export function rollingMean(arr: number[], window: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < arr.length; i++) {
    if (i < window - 1) { out.push(NaN); continue; }
    const slice = arr.slice(i - window + 1, i + 1);
    out.push(slice.reduce((a, b) => a + b, 0) / window);
  }
  return out;
}

export function rollingStd(arr: number[], window: number): number[] {
  const means = rollingMean(arr, window);
  const out: number[] = [];
  for (let i = 0; i < arr.length; i++) {
    if (i < window - 1) { out.push(NaN); continue; }
    const slice = arr.slice(i - window + 1, i + 1);
    const mean = means[i];
    const variance = slice.reduce((s, v) => s + (v - mean) ** 2, 0) / window;
    out.push(Math.sqrt(variance));
  }
  return out;
}

export function computeObv(bars: OHLCVBar[]): number[] {
  const obv = [0];
  for (let i = 1; i < bars.length; i++) {
    if (bars[i].close > bars[i - 1].close) obv.push(obv[i - 1] + bars[i].volume);
    else if (bars[i].close < bars[i - 1].close) obv.push(obv[i - 1] - bars[i].volume);
    else obv.push(obv[i - 1]);
  }
  return obv;
}

export function findSupportResistance(bars: OHLCVBar[], lookback = 60) {
  const recent = bars.slice(-lookback);
  // Use 10th / 90th percentile-ish via sorted extremes to reduce outlier stops
  const lows = [...recent.map((b) => b.low)].sort((a, b) => a - b);
  const highs = [...recent.map((b) => b.high)].sort((a, b) => a - b);
  const li = Math.max(0, Math.floor(lows.length * 0.1));
  const hi = Math.min(highs.length - 1, Math.floor(highs.length * 0.9));
  return {
    support: lows[li],
    resistance: highs[hi],
  };
}

export function detectTrend(bars: OHLCVBar[]): string {
  if (bars.length < 30) return "neutral";
  const closes = bars.map((b) => b.close);
  const sma = (n: number) => closes.slice(-n).reduce((a, b) => a + b, 0) / n;
  const sma20 = sma(20);
  const sma50 = bars.length >= 50 ? sma(50) : sma20;
  const price = closes[closes.length - 1];
  const highs = bars.slice(-10).map((b) => b.high);
  const lows = bars.slice(-10).map((b) => b.low);
  const count = (arr: number[], fn: (a: number, b: number) => boolean) =>
    arr.slice(1).filter((v, i) => fn(v, arr[i])).length;
  const hh = count(highs, (a, b) => a > b);
  const hl = count(lows, (a, b) => a > b);
  const lh = count(highs, (a, b) => a < b);
  const ll = count(lows, (a, b) => a < b);
  if (price > sma20 && sma20 > sma50 && hh >= 4 && hl >= 4) return "uptrend";
  if (price < sma20 && sma20 < sma50 && lh >= 4 && ll >= 4) return "downtrend";
  return "neutral";
}

export function computeAtr(bars: OHLCVBar[], period = 14): number {
  if (bars.length < period + 1) {
    const closes = bars.map((b) => b.close);
    const mean = closes.reduce((a, b) => a + b, 0) / closes.length;
    const variance = closes.reduce((s, v) => s + (v - mean) ** 2, 0) / closes.length;
    return Math.sqrt(variance) * 0.02;
  }
  const trs: number[] = [];
  for (let i = 1; i < bars.length; i++) {
    const h = bars[i].high, l = bars[i].low, pc = bars[i - 1].close;
    trs.push(Math.max(h - l, Math.abs(h - pc), Math.abs(l - pc)));
  }
  // Wilder smoothing for ATR
  let atr = trs.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (let i = period; i < trs.length; i++) {
    atr = (atr * (period - 1) + trs[i]) / period;
  }
  return atr;
}

/** True Wilder RSI(14) */
export function computeRsi(bars: OHLCVBar[], period = 14): number {
  if (bars.length < period + 1) return 50;
  const closes = bars.map((b) => b.close);
  let avgGain = 0;
  let avgLoss = 0;
  for (let i = 1; i <= period; i++) {
    const diff = closes[i] - closes[i - 1];
    if (diff >= 0) avgGain += diff;
    else avgLoss -= diff;
  }
  avgGain /= period;
  avgLoss /= period;
  for (let i = period + 1; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1];
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
  }
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return Math.round((100 - 100 / (1 + rs)) * 10) / 10;
}

export function mvrbMetrics(bars: OHLCVBar[], niftyBars: OHLCVBar[]) {
  if (bars.length < 22) return null;
  const price = bars[bars.length - 1].close;
  const price1m = bars[bars.length - 22].close;
  const lookback3m = Math.min(bars.length - 1, 66);
  const price3m = bars[bars.length - 1 - lookback3m].close;
  const ret1m = ((price - price1m) / price1m) * 100;
  const ret3m = ((price - price3m) / price3m) * 100;
  const vols = bars.map((b) => b.volume);
  const volAvg = rollingMean(vols, 20).at(-1) || 1;
  const volRatio = bars[bars.length - 1].volume / volAvg;
  const maxPrice = Math.max(...bars.slice(-252).map((b) => b.close));
  let rs = 1;
  if (niftyBars.length >= lookback3m + 1) {
    const n0 = niftyBars[niftyBars.length - 1 - lookback3m].close;
    const n1 = niftyBars.at(-1)!.close;
    const niftyRet = ((n1 - n0) / n0) * 100;
    // Relative strength as excess return (stable when index ≈ flat)
    if (Math.abs(niftyRet) < 0.5) {
      rs = 1 + ret3m / 10;
    } else {
      rs = 1 + (ret3m - niftyRet) / Math.max(5, Math.abs(niftyRet));
    }
  }
  return {
    ret_1m: Math.round(ret1m * 100) / 100,
    ret_3m: Math.round(ret3m * 100) / 100,
    vol_ratio: Math.round(volRatio * 100) / 100,
    near_52w_high: price / maxPrice > 0.95,
    rs_vs_nifty: Math.round(rs * 100) / 100,
  };
}

export function accumulationBreakout(bars: OHLCVBar[]) {
  if (bars.length < 30) {
    return { signal: false, price_consolidation: false, volume_accumulation: false, near_20sma: false, pre_breakout: false };
  }
  const closes = bars.map((b) => b.close);
  const vols = bars.map((b) => b.volume);
  const sma20 = rollingMean(closes, 20);
  const volSma = rollingMean(vols, 20);
  const std20 = rollingStd(closes, 20);
  const i = bars.length - 1;
  const bbWidth = 4 * (std20[i] || 0);
  const bbWidths = closes.map((_, idx) => 4 * (std20[idx] || 0));
  const bbWidthAvg = rollingMean(bbWidths, 20).at(-1) || bbWidth;
  const priceConsolidation = bbWidth < bbWidthAvg * 0.75;
  const volumeAccum = vols[i] > (volSma[i] || 0) * 1.5;
  const near20sma = Math.abs(closes[i] - (sma20[i] || closes[i])) / (sma20[i] || closes[i]) < 0.03;
  const notAtHigh = closes[i] < Math.max(...closes) * 0.95;
  const priceChg15d = closes[i] - closes[Math.max(0, i - 15)];
  const priceChgPct = (priceChg15d / closes[Math.max(0, i - 15)]) * 100;
  return {
    signal: priceConsolidation && volumeAccum && near20sma && notAtHigh,
    price_consolidation: priceConsolidation,
    volume_accumulation: volumeAccum,
    near_20sma: near20sma,
    pre_breakout: notAtHigh,
    bb_width: Math.round(bbWidth * 100) / 100,
    vol_vs_avg: Math.round((vols[i] / (volSma[i] || 1)) * 100) / 100,
    price_change_15d_pct: Math.round(priceChgPct * 100) / 100,
  };
}

export function obvAccumulation(bars: OHLCVBar[]) {
  if (bars.length < 60) {
    return { signal: false, obv_divergence: false, volume_spike: false, price_change_15d: 0, obv_change_15d: 0, price_down_flat: false };
  }
  const obv = computeObv(bars);
  const priceChange = bars.at(-1)!.close - bars.at(-15)!.close;
  const obvChange = obv.at(-1)! - obv.at(-15)!;
  const priceChgPct = (priceChange / bars.at(-15)!.close) * 100;
  const avg20 = bars.slice(-20).reduce((s, b) => s + b.volume, 0) / 20;
  const avg60 = bars.slice(-60).reduce((s, b) => s + b.volume, 0) / 60;
  const volumeSpike = avg20 > 1.5 * avg60;
  const obvDivergence = obvChange > 0 && priceChange <= 0;
  const priceDownFlat = priceChgPct <= 2 && priceChgPct >= -8;
  // True accumulation requires divergence + flat/down price (not volume spike alone)
  return {
    signal: obvDivergence && priceDownFlat,
    obv_divergence: obvDivergence,
    volume_spike: volumeSpike,
    price_change_15d: Math.round(priceChange * 100) / 100,
    price_change_15d_pct: Math.round(priceChgPct * 100) / 100,
    obv_change_15d: Math.round(obvChange),
    price_down_flat: priceDownFlat,
    vol_20d_vs_60d: Math.round((avg20 / avg60) * 100) / 100,
  };
}

export function computeTechnicalScores(bars: OHLCVBar[], niftyBars: OHLCVBar[]) {
  const mvrb = mvrbMetrics(bars, niftyBars);
  const accum = accumulationBreakout(bars);
  const obv = obvAccumulation(bars);
  const trend = detectTrend(bars);
  const { support, resistance } = findSupportResistance(bars);
  const atr = computeAtr(bars);
  const rsi = computeRsi(bars);
  const price = bars.at(-1)!.close;
  const confluence = evaluateStrategyConfluence(bars);

  let trendScore = 50;
  if (trend === "uptrend") trendScore = 80;
  else if (trend === "downtrend") trendScore = 25;
  if (confluence.ema_stack_bull && trend === "uptrend") trendScore = Math.min(100, trendScore + 5);
  if (confluence.adx.trending && confluence.adx.plus_di > confluence.adx.minus_di) {
    trendScore = Math.min(100, trendScore + 5);
  }

  let momentumScore = 50;
  if (mvrb) {
    if (mvrb.ret_3m > 15) momentumScore = 85;
    else if (mvrb.ret_3m > 10) momentumScore = 75;
    else if (mvrb.ret_3m > 5) momentumScore = 65;
    else if (mvrb.ret_3m < -10) momentumScore = 20;
    else if (mvrb.ret_3m < -5) momentumScore = 30;
  }
  // RSI: constructive mid-trend is good; overbought is a penalty for new Buys
  if (rsi >= 50 && rsi <= 65) momentumScore = Math.min(100, momentumScore + 8);
  else if (rsi > 70) momentumScore = Math.max(0, momentumScore - 8);
  else if (rsi < 30) momentumScore = Math.max(0, momentumScore - 12);
  else if (rsi < 45) momentumScore = Math.max(0, momentumScore - 5);
  if (confluence.macd.above_signal && confluence.macd.hist_rising) {
    momentumScore = Math.min(100, momentumScore + 6);
  } else if (confluence.macd.bearish_cross) {
    momentumScore = Math.max(0, momentumScore - 8);
  }

  let volumeScore = 50;
  const volRatio = mvrb?.vol_ratio ?? 0;
  if (volRatio > 1.5) volumeScore = 85;
  else if (volRatio > 1.2) volumeScore = 70;
  else if (volRatio > 1.0) volumeScore = 60;
  else if (mvrb && volRatio < 0.7) volumeScore = 35;

  let srScore = 50;
  const distRes = ((resistance - price) / price) * 100;
  const distSup = ((price - support) / price) * 100;
  if (distRes < 2 && trend === "uptrend") srScore = 75;
  else if (distSup < 3 && trend === "uptrend") srScore = 70;
  if (confluence.bollinger.overbought) srScore = Math.max(0, srScore - 10);

  let rsScore = 50;
  if (mvrb) {
    if (mvrb.rs_vs_nifty > 1.2) rsScore = 85;
    else if (mvrb.rs_vs_nifty > 1.0) rsScore = 75;
    else if (mvrb.rs_vs_nifty > 0.9) rsScore = 65;
    else if (mvrb.rs_vs_nifty < 0.8) rsScore = 35;
  }

  // Bonus only for confirmed setups (OBV true divergence, not volume spike alone)
  let bonus = (accum.signal ? 10 : 0) + (obv.signal ? 8 : 0);
  if (mvrb?.near_52w_high && volRatio >= 1.2) bonus += 4; // chase only with volume
  if (confluence.bullish_confluence) bonus += 5;
  bonus = Math.min(bonus, 15);

  const technicalScore = Math.min(100, Math.max(0, Math.round(
    trendScore * 0.25 + momentumScore * 0.2 + volumeScore * 0.2 + srScore * 0.15 + rsScore * 0.2 + bonus * 0.6
  )));

  let entryLow: number, entryHigh: number, stopLoss: number, target1: number, target2: number;
  if (trend === "uptrend" && price > support) {
    entryLow = Math.round((price - atr * 0.3) * 100) / 100;
    entryHigh = Math.round((price + atr * 0.1) * 100) / 100;
    stopLoss = Math.round(Math.max(support - atr * 0.3, price - atr * 1.5) * 100) / 100;
    target1 = Math.round(Math.max(resistance, price + atr * 1.5) * 100) / 100;
    target2 = Math.round((price + atr * 2.5) * 100) / 100;
  } else if (trend === "downtrend") {
    entryLow = Math.round((price - atr * 0.2) * 100) / 100;
    entryHigh = price;
    stopLoss = Math.round(Math.min(resistance + atr * 0.3, price + atr * 1.5) * 100) / 100;
    target1 = Math.round(Math.min(support, price - atr * 1.5) * 100) / 100;
    target2 = Math.round((price - atr * 2.5) * 100) / 100;
  } else {
    entryLow = Math.round(support * 100) / 100;
    entryHigh = price;
    stopLoss = Math.round((support - atr * 0.5) * 100) / 100;
    target1 = Math.round(resistance * 100) / 100;
    target2 = Math.round(((price + resistance) / 2) * 100) / 100;
  }

  const risk = Math.abs(price - stopLoss);
  const reward = Math.abs(target1 - price);
  const signals: string[] = [];
  if (trend === "uptrend") signals.push("Price in uptrend with higher highs/lows");
  else if (trend === "downtrend") signals.push("Price in downtrend");
  if (rsi >= 50 && rsi <= 65) signals.push(`RSI constructive at ${rsi}`);
  else if (rsi > 70) signals.push(`RSI overbought at ${rsi} — avoid fresh longs`);
  else if (rsi < 30) signals.push(`RSI oversold at ${rsi}`);
  if (mvrb && mvrb.ret_3m > 10) signals.push(`Strong 3M momentum: ${mvrb.ret_3m}%`);
  if (mvrb && mvrb.vol_ratio > 1.2) signals.push(`Volume above average: ${mvrb.vol_ratio}x`);
  if (mvrb && mvrb.rs_vs_nifty > 1) signals.push(`Outperforming Nifty (RS: ${mvrb.rs_vs_nifty})`);
  if (mvrb?.near_52w_high) signals.push("Trading near 52-week high");
  if (accum.signal) signals.push("Consolidation with volume accumulation — potential breakout");
  if (obv.obv_divergence) {
    signals.push(
      `OBV rising (+${obv.obv_change_15d}) while price ${(obv.price_change_15d_pct ?? 0) <= 0 ? "down" : "flat"} (${obv.price_change_15d_pct ?? 0}%) — quiet accumulation`,
    );
  }
  if (obv.volume_spike && !obv.obv_divergence) {
    signals.push(`Volume spike: 20-day avg is ${obv.vol_20d_vs_60d}x of 60-day avg (needs price/OBV confirm)`);
  }
  for (const f of confluence.factors.slice(0, 4)) signals.push(f);

  const activeSetups: string[] = [];
  if (obv.obv_divergence && obv.price_down_flat) activeSetups.push("obv_accumulation");
  if (accum.signal) activeSetups.push("volume_breakout_setup");
  // Stricter MVRB setup: need real volume expansion
  if (mvrb && mvrb.ret_3m > 12 && mvrb.vol_ratio >= 1.3 && mvrb.rs_vs_nifty >= 1.0) {
    activeSetups.push("mvrb_momentum");
  }
  if (confluence.bullish_confluence && confluence.macd.bullish_cross) activeSetups.push("macd_cross");
  if (confluence.bullish_confluence && confluence.bollinger.squeeze) activeSetups.push("bb_squeeze_breakout");
  if (confluence.adx.trending && confluence.ema_stack_bull) activeSetups.push("adx_ema_trend");

  return {
    technical_score: technicalScore,
    trend,
    trend_score: trendScore,
    momentum_score: momentumScore,
    volume_score: volumeScore,
    sr_score: srScore,
    rs_score: rsScore,
    rsi,
    mvrb,
    accumulation: accum,
    obv,
    strategy_confluence: confluence,
    active_setups: activeSetups,
    support,
    resistance,
    atr: Math.round(atr * 100) / 100,
    current_price: price,
    entry_zone: [entryLow, entryHigh] as [number, number],
    stop_loss: stopLoss,
    target1,
    target2,
    risk_reward: risk > 0 ? Math.round((reward / risk) * 100) / 100 : 0,
    technical_signals: signals,
    invalidation: trend !== "downtrend" ? `Close below ₹${stopLoss}` : `Close above ₹${stopLoss}`,
  };
}

File: backend/app/engines/technical.py
from typing import Any

import numpy as np
import pandas as pd


def compute_obv(df: pd.DataFrame) -> list[float]:
    obv = [0.0]
    for i in range(1, len(df)):
        if df["close"].iloc[i] > df["close"].iloc[i - 1]:
            obv.append(obv[-1] + df["volume"].iloc[i])
        elif df["close"].iloc[i] < df["close"].iloc[i - 1]:
            obv.append(obv[-1] - df["volume"].iloc[i])
        else:
            obv.append(obv[-1])
    return obv


def find_support_resistance(df: pd.DataFrame, lookback: int = 60) -> tuple[float, float]:
    recent = df.tail(lookback)
    support = recent["low"].min()
    resistance = recent["high"].max()
    return float(support), float(resistance)


def detect_trend(df: pd.DataFrame) -> str:
    if len(df) < 30:
        return "neutral"
    closes = df["close"].tail(20)
    sma20 = df["close"].rolling(20).mean().iloc[-1]
    sma50 = df["close"].rolling(50).mean().iloc[-1] if len(df) >= 50 else sma20
    price = df["close"].iloc[-1]

    highs = df["high"].tail(10).values
    lows = df["low"].tail(10).values
    hh = sum(1 for i in range(1, len(highs)) if highs[i] > highs[i - 1])
    hl = sum(1 for i in range(1, len(lows)) if lows[i] > lows[i - 1])
    lh = sum(1 for i in range(1, len(highs)) if highs[i] < highs[i - 1])
    ll = sum(1 for i in range(1, len(lows)) if lows[i] < lows[i - 1])

    if price > sma20 > sma50 and hh >= 4 and hl >= 4:
        return "uptrend"
    if price < sma20 < sma50 and lh >= 4 and ll >= 4:
        return "downtrend"
    return "neutral"


def compute_atr(df: pd.DataFrame, period: int = 14) -> float:
    if len(df) < period + 1:
        return df["close"].std() * 0.02 if len(df) > 1 else 0
    high, low, close = df["high"], df["low"], df["close"]
    tr = pd.concat(
        [
            high - low,
            (high - close.shift()).abs(),
            (low - close.shift()).abs(),
        ],
        axis=1,
    ).max(axis=1)
    return float(tr.rolling(period).mean().iloc[-1])


def mvrb_metrics(df: pd.DataFrame, nifty_df: pd.DataFrame) -> dict[str, Any]:
    """MVRB strategy from mvsp.py: momentum, volume, relative strength, breakout."""
    if len(df) < 22:
        return {}

    price_today = df["close"].iloc[-1]
    price_1m = df["close"].iloc[-22]
    price_3m = df["close"].iloc[0] if len(df) >= 60 else df["close"].iloc[0]

    ret_1m = (price_today - price_1m) / price_1m * 100
    ret_3m = (price_today - price_3m) / price_3m * 100

    vol_today = df["volume"].iloc[-1]
    vol_avg = df["volume"].rolling(20).mean().iloc[-1]
    vol_ratio = vol_today / vol_avg if vol_avg else 0

    max_price = df["close"].max()
    near_breakout = (price_today / max_price) > 0.95 if max_price else False

    rs_score = 1.0
    if not nifty_df.empty and len(nifty_df) >= 2:
        nifty_ret = (nifty_df["close"].iloc[-1] - nifty_df["close"].iloc[0]) / nifty_df["close"].iloc[0] * 100
        rs_score = ret_3m / nifty_ret if nifty_ret else 1.0

    return {
        "ret_1m": round(ret_1m, 2),
        "ret_3m": round(ret_3m, 2),
        "vol_ratio": round(vol_ratio, 2),
        "near_52w_high": near_breakout,
        "rs_vs_nifty": round(rs_score, 2),
    }


def accumulation_breakout(df: pd.DataFrame) -> dict[str, Any]:
    """BB width consolidation + volume accumulation from mvsp.py."""
    if len(df) < 30:
        return {"signal": False}

    df = df.copy()
    df["close_20sma"] = df["close"].rolling(20).mean()
    df["volume_20sma"] = df["volume"].rolling(20).mean()
    std20 = df["close"].rolling(20).std()
    df["bb_upper"] = df["close_20sma"] + 2 * std20
    df["bb_lower"] = df["close_20sma"] - 2 * std20
    df["bb_width"] = df["bb_upper"] - df["bb_lower"]

    latest = df.iloc[-1]
    if latest.isnull().any():
        return {"signal": False}

    bb_width_avg = df["bb_width"].rolling(20).mean().iloc[-1]
    price_consolidation = latest["bb_width"] < bb_width_avg * 0.75 if bb_width_avg else False
    volume_accum = latest["volume"] > latest["volume_20sma"] * 1.3
    near_20sma = abs(latest["close"] - latest["close_20sma"]) / latest["close_20sma"] < 0.03
    not_at_high = latest["close"] < df["close"].max() * 0.95

    signal = price_consolidation and volume_accum and near_20sma and not_at_high
    return {
        "signal": bool(signal),
        "price_consolidation": price_consolidation,
        "volume_accumulation": volume_accum,
        "near_20sma": near_20sma,
        "pre_breakout": not_at_high,
        "bb_width": round(float(latest["bb_width"]), 2),
    }


def obv_accumulation(df: pd.DataFrame) -> dict[str, Any]:
    """OBV divergence + volume spike from mvsp.py."""
    if len(df) < 60:
        return {"signal": False}

    df = df.copy()
    df["obv"] = compute_obv(df)

    price_change = df["close"].iloc[-1] - df["close"].iloc[-15]
    obv_change = df["obv"].iloc[-1] - df["obv"].iloc[-15]

    avg_vol_20 = df["volume"].iloc[-20:].mean()
    avg_vol_60 = df["volume"].iloc[-60:].mean()
    volume_spike = avg_vol_20 > 1.5 * avg_vol_60

    obv_divergence = obv_change > 0 and price_change <= 0
    near_52w = df["close"].iloc[-1] >= 0.9 * df["close"].max()

    signal = volume_spike or obv_divergence
    return {
        "signal": bool(signal),
        "obv_divergence": obv_divergence,
        "volume_spike": volume_spike,
        "near_52w_high": near_52w,
    }


def compute_technical_scores(
    df: pd.DataFrame,
    nifty_df: pd.DataFrame,
) -> dict[str, Any]:
    mvrb = mvrb_metrics(df, nifty_df)
    accum = accumulation_breakout(df)
    obv = obv_accumulation(df)
    trend = detect_trend(df)
    support, resistance = find_support_resistance(df)
    atr = compute_atr(df)
    price = float(df["close"].iloc[-1])

    trend_score = 50
    if trend == "uptrend":
        trend_score = 80
    elif trend == "downtrend":
        trend_score = 25

    momentum_score = 50
    if mvrb:
        if mvrb.get("ret_3m", 0) > 15:
            momentum_score = 85
        elif mvrb.get("ret_3m", 0) > 10:
            momentum_score = 75
        elif mvrb.get("ret_3m", 0) > 5:
            momentum_score = 65
        elif mvrb.get("ret_3m", 0) < -10:
            momentum_score = 20
        elif mvrb.get("ret_3m", 0) < -5:
            momentum_score = 30

    volume_score = 50
    if mvrb and mvrb.get("vol_ratio", 0) > 1.5:
        volume_score = 85
    elif mvrb and mvrb.get("vol_ratio", 0) > 1.2:
        volume_score = 70
    elif mvrb and mvrb.get("vol_ratio", 0) > 1.0:
        volume_score = 60
    elif mvrb and mvrb.get("vol_ratio", 0) < 0.7:
        volume_score = 35

    sr_score = 50
    dist_to_res = (resistance - price) / price * 100
    dist_to_sup = (price - support) / price * 100
    if dist_to_res < 2 and trend == "uptrend":
        sr_score = 75
    elif dist_to_sup < 3 and trend == "uptrend":
        sr_score = 70
    elif dist_to_res < 1:
        sr_score = 65

    rs_score_val = 50
    if mvrb:
        rs = mvrb.get("rs_vs_nifty", 1)
        if rs > 1.2:
            rs_score_val = 85
        elif rs > 1.0:
            rs_score_val = 75
        elif rs > 0.9:
            rs_score_val = 65
        elif rs < 0.8:
            rs_score_val = 35

    breakout_bonus = 10 if accum.get("signal") else 0
    obv_bonus = 8 if obv.get("signal") else 0
    breakout_flag = mvrb.get("near_52w_high", False) if mvrb else False
    breakout_bonus += 5 if breakout_flag else 0

    raw_technical = (
        trend_score * 0.25
        + momentum_score * 0.2
        + volume_score * 0.15
        + sr_score * 0.15
        + rs_score_val * 0.15
        + min(breakout_bonus + obv_bonus, 15)
    )
    technical_score = min(100, max(0, round(raw_technical)))

    if trend == "uptrend" and price > support:
        entry_low = round(price - atr * 0.3, 2)
        entry_high = round(price + atr * 0.1, 2)
        stop_loss = round(support - atr * 0.5, 2)
        target1 = round(resistance, 2)
        target2 = round(price + (resistance - price) * 1.5, 2) if resistance > price else round(price + atr * 2, 2)
    elif trend == "downtrend":
        entry_low = round(price - atr * 0.2, 2)
        entry_high = round(price, 2)
        stop_loss = round(resistance + atr * 0.3, 2)
        target1 = round(support, 2)
        target2 = round(support - atr, 2)
    else:
        entry_low = round(support, 2)
        entry_high = round(price, 2)
        stop_loss = round(support - atr * 0.5, 2)
        target1 = round(resistance, 2)
        target2 = round((price + resistance) / 2, 2)

    risk = abs(price - stop_loss)
    reward = abs(target1 - price)
    rr_ratio = round(reward / risk, 2) if risk > 0 else 0

    signals = []
    if trend == "uptrend":
        signals.append("Price in uptrend with higher highs/lows")
    elif trend == "downtrend":
        signals.append("Price in downtrend")
    if mvrb:
        if mvrb.get("ret_3m", 0) > 10:
            signals.append(f"Strong 3M momentum: {mvrb['ret_3m']}%")
        if mvrb.get("vol_ratio", 0) > 1.2:
            signals.append(f"Volume above average: {mvrb['vol_ratio']}x")
        if mvrb.get("rs_vs_nifty", 0) > 1:
            signals.append(f"Outperforming Nifty (RS: {mvrb['rs_vs_nifty']})")
        if mvrb.get("near_52w_high"):
            signals.append("Trading near 52-week high")
    if accum.get("signal"):
        signals.append("Consolidation with volume accumulation — potential breakout")
    if obv.get("obv_divergence"):
        signals.append("OBV rising while price flat — accumulation signal")
    if obv.get("volume_spike"):
        signals.append("Recent volume spike vs 60-day average")

    return {
        "technical_score": technical_score,
        "trend": trend,
        "trend_score": trend_score,
        "momentum_score": momentum_score,
        "volume_score": volume_score,
        "sr_score": sr_score,
        "rs_score": rs_score_val,
        "mvrb": mvrb,
        "accumulation": accum,
        "obv": obv,
        "support": support,
        "resistance": resistance,
        "atr": round(atr, 2),
        "current_price": price,
        "entry_zone": [entry_low, entry_high],
        "stop_loss": stop_loss,
        "target1": target1,
        "target2": target2,
        "risk_reward": rr_ratio,
        "technical_signals": signals,
        "invalidation": f"Close below ₹{stop_loss}" if trend != "downtrend" else f"Close above ₹{stop_loss}",
    }
